xwayland/xwayland.c

```diff
@@ -323,36 +351,74 @@ static bool xwayland_start(struct wlr_xwayland *wlr_xwayland,
 	}
 	if (wlr_xwayland->pid < 0) {
 		wlr_log_errno(L_ERROR, "fork failed");
-		xwayland_finish(wlr_xwayland);
+		xwayland_finish_server(wlr_xwayland);
 		return false;
 	}
 
 	/* close child fds */
-	close(wlr_xwayland->x_fd[0]);
-	close(wlr_xwayland->x_fd[1]);
+	/* remain managing x sockets for lazy start */
 	close(wlr_xwayland->wl_fd[1]);
 	close(wlr_xwayland->wm_fd[1]);
-	wlr_xwayland->x_fd[0] = wlr_xwayland->x_fd[1] = -1;
 	wlr_xwayland->wl_fd[1] = wlr_xwayland->wm_fd[1] = -1;
 
 	return true;
 }
 
+static bool xwayland_start_server_lazy(struct wlr_xwayland *wlr_xwayland) {
+	struct wl_event_loop *loop = wl_display_get_event_loop(wlr_xwayland->wl_display);
+	wlr_xwayland->x_fd_read_event[0] =
+		wl_event_loop_add_fd(loop, wlr_xwayland->x_fd[0], WL_EVENT_READABLE,
+				xwayland_socket_connected, wlr_xwayland);
+	wlr_xwayland->x_fd_read_event[1] =
+		wl_event_loop_add_fd(loop, wlr_xwayland->x_fd[1], WL_EVENT_READABLE,
+				xwayland_socket_connected, wlr_xwayland);
+
+	return true;
+}
+
 void wlr_xwayland_destroy(struct wlr_xwayland *wlr_xwayland) {
 	wlr_xwayland_set_seat(wlr_xwayland, NULL);
-	xwayland_finish(wlr_xwayland);
+	xwayland_finish_server(wlr_xwayland);
+	xwayland_finish_display(wlr_xwayland);
 	free(wlr_xwayland);
 }
 
 struct wlr_xwayland *wlr_xwayland_create(struct wl_display *wl_display,
-		struct wlr_compositor *compositor) {
+		struct wlr_compositor *compositor, bool lazy) {
 	struct wlr_xwayland *wlr_xwayland = calloc(1, sizeof(struct wlr_xwayland));
+	memset(wlr_xwayland, 0, offsetof(struct wlr_xwayland, seat));
```

Comment:
That `memset` is unseless here because it's right after the `calloc` call. It's useful to reset the xwayland state (for instance during a restart). Should probably be moved in one of the `xwayland_start_*` functions.

Reply:
Additionally, the `calloc` call is missing a NULL check.